src/qibolab/unrolling.py

```diff
@@ -0,0 +1,79 @@
+"""Utilities for sequence unrolling.
+
+May be reused by different instruments.
+"""
+
+from dataclasses import asdict, dataclass, field, fields
+
+from .pulses import PulseSequence
+
+
+def _waveform(sequence: PulseSequence):
+    # TODO: deduplicate pulses (Not yet as drivers may not support it yet)
+    # TODO: count Rectangular and delays separately (Zurich Instruments supports this)
+    # TODO: Any constant part of a pulse should be counted only once (Zurich Instruments supports this)
+    # TODO: check if readout duration is faithful for the readout pulse (I would only check the control pulses)
```

Comment:
I'm not sure I see the reason for counting only control pulses...

Reply:
Because for what I saw on the device the problem was filling the control pulse channel memory as they are way more control pulses (drive usually) than readouts or anything else. Also, readouts could be like 50 control pulses in duration but they are compressible as they have constant parts and won't have the same impact that as many control pulses duration wise.

Reply:
@Jacfomg are you sure you are not reuploading the same pulse multiple times? Eg. in flipping you may have 100 pulses but they are all the same (pi pulse) so you only need to upload once. But okay, even if you upload once, maybe the instrument is generating multiple times internally, I am not sure.

For the constant part of the readout, it souds right.

Reply:
I'm quite sure I'm uploading them several times as the driver doesn't check for this kind of things.

Reply:
Even QUA is uploading infinite times if you do not deduplicate them explicitly, but there should be for sure a way to reuse existing waveforms on the device.

Reply:
I wasn't planning to look into that until @hay-k finishes his refactor

Reply:
Ok, however deduplicating in Zurich is simple (I just need to look up in our driver):

> All pulses accept the following keyword arguments: - uid ([str](https://docs.python.org/3/library/stdtypes.html#str)): Unique identifier of the pulse

https://docs.zhinst.com/labone_q_user_manual/reference/dsl/experiment/?h=pulse_library#laboneq.dsl.experiment.pulse_library

So, it's exactly as for @stavros11 in QUA.